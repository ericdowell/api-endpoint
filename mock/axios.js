export function BasicMock(options) {
  return new Promise((resolve, reject) => {
    if (options.errors) {
      return reject({
        config: {},
        response: {
          data: {
            errors: options.errors
          },
          status: options.status || 422
        }
      })
    }
    return resolve({
      config: {},
      data: { foo: 'bar' }
    })
  })
}
