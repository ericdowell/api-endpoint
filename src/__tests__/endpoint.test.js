/* global test, expect, jest */
import axios from 'axios'
import Endpoint from '../endpoint'
import { BasicMock } from '../../mock/axios'
import MessageBag from '../errors/messageBag'

jest.mock('axios')
jest.spyOn(global.console, 'log')

axios.mockImplementation(BasicMock)

test('the origin is localhost', () => {
  expect(new Endpoint().origin).toMatchSnapshot()
})

test('the path is empty', () => {
  expect(new Endpoint().path).toStrictEqual('')
})

test('the endpoint is empty', () => {
  expect(new Endpoint().endpoint).toStrictEqual('')
})

test('the getBaseUrl is localhost with trailing slash', () => {
  expect(new Endpoint().getBaseUrl()).toMatchSnapshot()
})

test('the headers default is empty', () => {
  expect(new Endpoint().hasHeaders()).toMatchSnapshot()
  expect(new Endpoint().headers).toMatchSnapshot()
})

test('the setHeaders updates headers value', () => {
  const endpoint = new Endpoint().setHeader('Accept', 'application/json')
  expect(endpoint.hasHeaders()).toMatchSnapshot()
  expect(endpoint.headers).toMatchSnapshot()
})

test('the queryOptions will return default options', () => {
  const options = new Endpoint().queryOptions('final/path', 'get')
  expect(options).toMatchSnapshot()
})

test('the responseData will return data key of object', () => {
  const data = new Endpoint().responseData({ data: 'test' })
  expect(data).toMatchSnapshot()
})

test('the query will return response object', done => {
  new Endpoint().query('123', 'get').then(response => {
    expect(response).toMatchSnapshot()
    done()
  })
})

test('the query will return MessageBag object', done => {
  new Endpoint()
    .query('123', 'get', { errors: { foo: 'bar is something else.' } })
    .catch(messageBag => {
      expect(messageBag).toBeInstanceOf(MessageBag)
      expect(messageBag.toString()).toMatchSnapshot()
      expect(messageBag.hasMessage('foo')).toMatchSnapshot()
      expect(messageBag.getMessage('foo')).toMatchSnapshot()
      expect(console.log).not.toHaveBeenCalled()
      done()
    })
})

test('the query will return MessageBag object and debug to console', done => {
  const endpoint = new Endpoint().debug()
  expect(endpoint.isDebugEnabled()).toMatchSnapshot()
  endpoint
    .query('123', 'get', { errors: { foo: 'bar is something else.' } })
    .catch(messageBag => {
      expect(messageBag).toBeInstanceOf(MessageBag)
      expect(messageBag.toString()).toMatchSnapshot()
      expect(messageBag.hasMessage('foo')).toMatchSnapshot()
      expect(messageBag.getMessage('foo')).toMatchSnapshot()
      expect(console.log).toHaveBeenCalled()
      done()
    })
})

test('the responseError will return default MessageBag', () => {
  const endpoint = new Endpoint()
  const options = endpoint.queryOptions('final/path', 'get')
  const messageBag = endpoint.responseError(
    options.url,
    options.method,
    options,
    new Error('failure to foo bar')
  )

  expect(messageBag).toBeInstanceOf(MessageBag)
  expect(messageBag.message).not.toStrictEqual('failure to foo bar')
  expect(messageBag.message).toStrictEqual(messageBag.default)
  expect(messageBag.getStatus()).toStrictEqual(messageBag.defaultStatus)
  expect(messageBag.hasMessage('test')).toStrictEqual(false)

  expect(MessageBag.toString()).toMatchSnapshot()
  expect(messageBag.options).toMatchSnapshot()
  expect(messageBag.original.message).toMatchSnapshot()
  expect(messageBag.original.toString()).toMatchSnapshot()
  expect(messageBag.url).toMatchSnapshot()
  expect(messageBag.method).toMatchSnapshot()
  expect(messageBag.config).toMatchSnapshot()
  expect(messageBag.response).toMatchSnapshot()
})
