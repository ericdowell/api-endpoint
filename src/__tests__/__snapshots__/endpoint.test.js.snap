// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the getBaseUrl is localhost with trailing slash 1`] = `"http://localhost/"`;

exports[`the origin is localhost 1`] = `"http://localhost"`;

exports[`the query will return MessageBag object 1`] = `"Error: An unexpected error has occurred. Please try again."`;

exports[`the query will return MessageBag object 2`] = `true`;

exports[`the query will return MessageBag object 3`] = `"bar is something else."`;

exports[`the query will return response object 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`the queryOptions will return default options 1`] = `
Object {
  "baseURL": "http://localhost/",
  "method": "GET",
  "paramsSerializer": [Function],
  "url": "final/path",
}
`;

exports[`the responseData will return data key of object 1`] = `"test"`;

exports[`the responseError will return default MessageBag 1`] = `
"function MessageBag() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Error.call.apply(_Error, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_this), \\"_url\\", '');

    _defineProperty(_assertThisInitialized(_this), \\"_method\\", 'unkown');

    _defineProperty(_assertThisInitialized(_this), \\"_options\\", {});

    _defineProperty(_assertThisInitialized(_this), \\"_config\\", {});

    _defineProperty(_assertThisInitialized(_this), \\"_response\\", {});

    _defineProperty(_assertThisInitialized(_this), \\"_original\\", null);

    return _this;
  }"
`;

exports[`the responseError will return default MessageBag 2`] = `
Object {
  "baseURL": "http://localhost/",
  "method": "GET",
  "paramsSerializer": [Function],
  "url": "final/path",
}
`;

exports[`the responseError will return default MessageBag 3`] = `"failure to foo bar"`;

exports[`the responseError will return default MessageBag 4`] = `"Error: failure to foo bar"`;

exports[`the responseError will return default MessageBag 5`] = `"final/path"`;

exports[`the responseError will return default MessageBag 6`] = `"GET"`;
